src/csharp/Intel/Iced/Intel/Extended/ExtendedEncoder.cs

```diff
@@ -0,0 +1,102 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+using Iced.Intel.BlockEncoderInternal;
+
+namespace Iced.Intel {
+	public sealed partial class ExtendedEncoder {
+
+		readonly CodeWriter _writer;
+		ulong _internalRip;
+		Label _label;
+		readonly List<Label> _labels;
+		readonly List<Instruction> _instructions;
+
+		private ExtendedEncoder(CodeWriter writer, int bitness) {
+			Debug.Assert(bitness == 16 || bitness == 32 || bitness == 64);
+			if (writer is null)
+				ThrowHelper.ThrowArgumentNullException_writer();
+			Bitness = bitness;
+			_writer = writer;
+			_labels = new List<Label>();
+			_instructions = new List<Instruction>();
+			_label = CreateLabel();
+			PreferVex = true;
+			PreferBranchNear = true;
+		}
+		
+		public ulong BaseRip { get; set; }
+
+		public int Bitness { get; }
+		
+		public bool PreferVex { get; set; }
+		
+		public bool PreferBranchNear { get; set; }
+
+		public static ExtendedEncoder Create(int bitness, CodeWriter writer) {
+			if (writer == null) throw new ArgumentNullException(nameof(writer));
+			switch (bitness) {
+			case 16:
+			case 32:
+			case 64:
+				return new ExtendedEncoder(writer, bitness);
+			default:
+				throw new ArgumentOutOfRangeException(nameof(bitness));
+			}
+		}
+
+		public Label CreateLabel(string name = null) {
+			_internalRip++;
+			var label = new Label(name, _internalRip);
+			_labels.Add(label);
+			return label;
+		}
+
+		public Label CurrentLabel => _label;
+
+		public void label(Label label) {
```

Comment:
I know I used `label()` in my example, but it's probably better to reserve all lower case method names to mnemonics, so call this `Label()` instead.